[diag] move `otPlatDiagModeGet()` check to `ProcessCmd`

maciejbaczmanski opened the pull request:
All diagnostic commands (instead of `diag start`) should fail if device is not in diagnostic mode. Previously it was verified by each command's process method (with a missing check in `ProcessEcho` and `ProcessGpio`). This commit moves the check directly to `ProcessCmd` and cleans up redundant code.

canisLupus1313 commented on src/core/diags/factory_diags.cpp:
```diff
@@ -1112,6 +1095,8 @@ Error Diags::ProcessCmd(uint8_t aArgsLength, char *aArgs[])
         ExitNow();
     }
 
+    VerifyOrExit(otPlatDiagModeGet() || StringMatch(aArgs[0], "start"), error = kErrorInvalidState);
```
You cold add info here that diag is not enabled.

maciejbaczmanski replied:
good point, added

zhanglongxia reviewed (approved):
LGTM

abtink reviewed (approved):
Good optimization about checking once in `ProcessCmd()`. Thanks. 👍 

We can consider doing something similar with `AppendErrorResult(error)` calls (which seem to be present at the end of all `Process{Something}()` methods). We can move them to `ProcessCmd()`. Maybe in a future PR?

maciejbaczmanski commented:
> Good optimization about checking once in `ProcessCmd()`. Thanks. 👍
> 
> We can consider doing something similar with `AppendErrorResult(error)` calls (which seem to be present at the end of all `Process{Something}()` methods). We can move them to `ProcessCmd()`. Maybe in a future PR?

I've moved `AppendErrorResult(error)` to  `ProcessCmd()` as I had to fix CI anyways.
I've also noticed that (if I understand the code correctly :) ) we can remove `diag rcp start/stop`. These commands set `otPlatDiagModeSet` which is set by `diag start/stop` anyways and IMO these commands can add some confusion. For example in `posix-diag-rcp.exp` we use only `diag start/stop`, while not using `diag rcp start/stop` anywhere.

maciejbaczmanski commented:
> I've also noticed that (if I understand the code correctly :) ) we can remove `diag rcp start/stop` commands. These commands set `otPlatDiagModeSet` which is set by `diag start/stop` anyways and IMO they can introduce some confusion. For example in `posix-diag-rcp.exp` we use only `diag start/stop`, while not using `diag rcp start/stop` anywhere.

OK, just found out that diag commands sent to RCP don't need to have `rcp` prefix, so `processstart` and `processstop` implementation on RCP side should not be removed. Reverted the change. I still think that in documentation we should not suggest using `diag rcp start` and `diag rcp stop` commands as in tests we also always start diag on host as well?

zhanglongxia commented:
> > I've also noticed that (if I understand the code correctly :) ) we can remove `diag rcp start/stop` commands. These commands set `otPlatDiagModeSet` which is set by `diag start/stop` anyways and IMO they can introduce some confusion. For example in `posix-diag-rcp.exp` we use only `diag start/stop`, while not using `diag rcp start/stop` anywhere.
> 
> OK, just found out that diag commands sent to RCP don't need to have `rcp` prefix, so `processstart` and `processstop` implementation on RCP side should not be removed. Reverted the change. I still think that in documentation we should not suggest using `diag rcp start` and `diag rcp stop` commands as in tests we also always start diag on host as well?

The `diag rcp` commands are used for debugging.  The description of the `diag rcp` command mentions it: https://github.com/openthread/openthread/blob/main/src/core/diags/README.md#diag-rcp. So I think we should not remove the `diag rcp start/stop` commands from the README.md.

The comment in the code:

> // This `rcp` command is for debugging and testing only, building only when NDEBUG is not defined
    // so that it will be excluded from release build.

maciejbaczmanski commented:
> > > I've also noticed that (if I understand the code correctly :) ) we can remove `diag rcp start/stop` commands. These commands set `otPlatDiagModeSet` which is set by `diag start/stop` anyways and IMO they can introduce some confusion. For example in `posix-diag-rcp.exp` we use only `diag start/stop`, while not using `diag rcp start/stop` anywhere.
> > 
> > 
> > OK, just found out that diag commands sent to RCP don't need to have `rcp` prefix, so `processstart` and `processstop` implementation on RCP side should not be removed. Reverted the change. I still think that in documentation we should not suggest using `diag rcp start` and `diag rcp stop` commands as in tests we also always start diag on host as well?
> 
> The `diag rcp` commands are used for debugging. The description of the `diag rcp` command mentions it: https://github.com/openthread/openthread/blob/main/src/core/diags/README.md#diag-rcp. So I think we should not remove the `diag rcp start/stop` commands from the README.md.
> 
> The comment in the code:
> 
> > // This `rcp` command is for debugging and testing only, building only when NDEBUG is not defined
> > // so that it will be excluded from release build.

hmm ok, thanks, dropped removing it from README then

zhanglongxia reviewed (approved):
LGTM

abtink reviewed (approved):
LGTM. 👍

jwhui reviewed (approved):
Thanks! 👍🏼
